linkerd/app/core/src/control.rs

```diff
@@ -301,6 +301,211 @@ pub mod resolve {
+    pub struct Resolution {
+        current_targets: HashSet<Target>,
+        address: ControlAddr,
+        dns: dns::Resolver,
+        #[pin]
+        state: State,
+    }
+
+    #[pin_project(project = StateProj)]
+    enum State {
+        Resolving(#[pin] dns::ResolveResponseFuture),
+        Resolved(VecDeque<Update<Target>>, #[pin] time::Delay),
+        Constant(Option<Target>),
+    }
+
+    // === impl ControlPlaneResolve ===
+
+    impl Resolve {
+        pub fn new(dns: dns::Resolver) -> Self {
+            Self { dns }
+        }
+    }
+
+    impl tower::Service<ControlAddr> for Resolve {
+        type Response = Resolution;
+        type Error = Error;
+        type Future =
+            Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+        fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+            Poll::Ready(Ok(()))
+        }
+
+        fn call(&mut self, target: ControlAddr) -> Self::Future {
+            Box::pin(futures::future::ok(match target.addr {
+                Addr::Socket(sa) => Resolution {
+                    current_targets: HashSet::default(),
+                    address: target.clone(),
+                    dns: self.dns.clone(),
+                    state: State::Constant(Some(Target {
+                        addr: sa,
+                        server_name: target.identity.clone(),
+                    })),
+                },
+                Addr::Name(ref na) => Resolution {
+                    current_targets: HashSet::default(),
+                    address: target.clone(),
+                    dns: self.dns.clone(),
+                    state: State::Resolving(self.dns.resolve_ips(na.name())),
+                },
+            }))
+        }
+    }
+
+    impl Resolution {
+        pub fn diff_targets(
+            old_targets: &HashSet<Target>,
+            new_targets: &HashSet<Target>,
+        ) -> (Vec<(SocketAddr, Target)>, Vec<SocketAddr>) {
+            let mut adds = Vec::default();
+            let mut removes = Vec::default();
+
+            for new_target in new_targets.iter() {
+                if !old_targets.contains(new_target) {
+                    adds.push((new_target.addr, new_target.clone()))
+                }
+            }
+
+            for old_target in old_targets.iter() {
+                if !new_targets.contains(old_target) {
+                    removes.push(old_target.addr)
+                }
+            }
+
+            (adds, removes)
+        }
+    }
+
+    // === impl ControlPlaneResolution ===
+
+    impl resolve::Resolution for Resolution {
+        type Endpoint = Target;
+        type Error = Error;
+
+        fn poll(
+            self: Pin<&mut Self>,
+            cx: &mut Context<'_>,
+        ) -> Poll<Result<Update<Self::Endpoint>, Self::Error>> {
+            let mut this = self.project();
+            loop {
+                match this.state.as_mut().project() {
```

Comment:
It's really a bummer that we have to write this large manual state machine here rather than being able to use `async-stream`, but I suppose this is because we need a `Resolution` implementation rather than a `Stream`...

Here's a thought: what if we added a function `linkerd2_proxy_core::resolve::from_stream` that looked like this:
```rust
pub fn from_stream<S>(stream: S) -> FromStream<S> {
     FromStream(stream)
}

#[pin_project]
pub struct FromStream<S>(#[pin] S);

impl<E, S> Resolution for FromStream<S>
where
    S: TryStream<Ok = Update<E>>,
    S::Error: Into<Error>,
{
    type Endpoint = E;
    type Error = S::Error;

    fn poll(
        self: Pin<&mut Self>,
        cx: &mut Context<'_>,
    ) -> Poll<Result<Update<Self::Endpoint>, Self::Error>> {
        let res = futures::ready!(self.project().0.try_poll_next(cx))
            .expect("resolution stream must be infinite!");
        Poll::Ready(res)
    }
}
```

Then, we could implement `Resolution` here by implementing a `Stream` using `async-stream`, and turning that `Stream` into a `Resolution`. This avoids the manual state machine implementation (and would let us use unboxed futures for the DNS resolution)...